async-queue/src/fifo.rs

```diff
@@ -0,0 +1,403 @@
+        };
+        if prev == current {
+            break (new, was_empty);
+        }
+    };
+
+    // 4. Write the data, then the `id`.
+    let slot = unsafe { &mut *descriptor.data.add(new.write_offset()) };
+    slot.copy_except_id(&val);
+    slot.set_id(val.get_id());
+
+    // 5. If the queue was empty in step 1, signal the reader to wake up.
+    Ok(was_empty)
+}
+
+pub(crate) fn try_recv_impl<T: WithAtomicId>(descriptor: &FifoDescriptor<T>) -> Result<(T, /*wake up writer:*/ bool), TryRecvError> {
+    // 1. Load the current offsets.
+    let current = unsafe {
+        let offsets = (*descriptor.offsets).load(Ordering::SeqCst);
+        Offsets::new(offsets, descriptor.len as u32)
+    };
+    let was_full = current.is_full();
+
+    // 2. If the queue is empty, wait, then go to step 1.
+    if current.is_empty() {
+        return Err(TryRecvError::QueueEmpty);
+    }
+
+    // 3. Add 1 to the read offset.
+    let new = current.increment_read_offset();
+
+    let slot = loop {
+        // 4. Read the `id` at the new read offset.
+        let slot = unsafe { &mut *descriptor.data.add(new.read_offset()) };
+        let id = slot.get_id();
+
+        // 5. If `id` is `0`, go to step 4 (spin). Spinning is OK because data is
+        //    expected to be written imminently.
+        if id != 0 {
+            break slot;
+        }
+    };
+
+    // 6. Read the data, then store `0` in the `id`.
+    let mut val = T::empty();
+    val.copy_except_id(slot);
+    val.set_id(slot.get_id());
+    slot.set_id(0);
+
+    // 7. Store the new read offset.
+    let after = unsafe {
+        fetch_adjust(
+            &*descriptor.offsets,
+            new.read as isize - current.read as isize,
+            Ordering::SeqCst,
+        )
+    };
+
+    // 8. If the queue was full in step 1, signal the writer to wake up.
+    //    ... or became full during read
+    let became_full = Offsets::new(after, descriptor.len as u32).is_full();
+    Ok((val, was_full || became_full))
+}
+
+lazy_static! {
+    pub(crate) static ref RECEIVER_TRACKER: ReceiverTracker = ReceiverTracker::new();
+}
+
+pub(crate) struct ReceiverTracker(Mutex<HashSet<usize>>);
+
+impl ReceiverTracker {
+    fn new() -> Self {
+        Self(Mutex::new(HashSet::new()))
+    }
+
+    pub(crate) fn new_receiver(&self, data_ptr: usize) {
+        let already_exists = {
+            let mut receivers = self.0.lock().unwrap();
+            !receivers.insert(data_ptr)
+        };
+        if already_exists {
+            panic!("Multiple receivers for the same Fifo is not allowed.");
+        }
+    }
+
+    pub(crate) fn drop_receiver(&self, data_ptr: usize) {
+        let mut receivers = self.0.lock().unwrap();
+        receivers.remove(&data_ptr);
+    }
+}
+
+// Note: we cannot have an AtomicU64 id in Usercall/Return types since they
+// need to be Copy due to requirements of `UserSafeSized` (see definition of
```

Comment:
I don't think so? `UserSafeSized` just needs `FifoDescriptor<T>` to be `Copy`, but that doesn't require `T: Copy`. So it should be possible to just convert the `id` fields to `AtomicU64`? Even if not, I prefer type casting over transmutes (transmuting a reference is pretty pointless).

This is related to something else I noticed: it would be good if we could enforce some kind of `Copy` bound on `WithAtomicId`. We could perhaps rearchitect the whole thing as follows:

```
#[repr(C)]
struct WithId<T: Copy> {
    id: AtomicU64,
    data: T,
}
```

I have a faint memory that we used to have that, but I don't remember the details.

Reply:
I like the idea of re-architecting as you pointed out. I'll try to do that.

The reason for not being able to just change the id fields in `Usercall`/`Return` types should be apparent by looking at the following definitions:
```rust
pub unsafe trait UserSafeSized: Copy + Sized {}
//...
unsafe impl UserSafeSized for Usercall {}
unsafe impl UserSafeSized for Return {}
```